### PR TITLE
Merge background contexts of all ingester pushes into one in distributor

### PR DESCRIPTION
**Pull request by wgliang**
Signed-off-by: wangguoliang <iamwgliang@gmail.com>

**What this PR does**:
Merge background contexts of all ingester pushes into one in distributor.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -763,6 +763,15 @@ func (d *Distributor) Push(ctx context.Context, req *cortexpb.WriteRequest) (*co
...
 
+	// Use a background context to make sure all ingesters get samples even if we return early
+	localCtx, cancel := context.WithTimeout(context.Background(), d.cfg.RemoteTimeout)
+	localCtx = user.InjectOrgID(localCtx, userID)
+	// Get clientIP(s) from Context and add it to localCtx
+	localCtx = util.AddSourceIPsToOutgoingContext(localCtx, source)
+	if sp := opentracing.SpanFromContext(ctx); sp != nil {
+		localCtx = opentracing.ContextWithSpan(localCtx, sp)
+	}
+
```
Comment (alanprot) [inline pkg/distributor/distributor.go:774]: We are missing this?

```
		if headerMap := util_log.HeaderMapFromContext(ctx); headerMap != nil {
			localCtx = util_log.ContextWithHeaderMap(localCtx, headerMap)
		}
```

Comment (wgliang) [reply pkg/distributor/distributor.go:774]: nice catch

**Review by alanprot — APPROVED** [review]

**Comment by alanprot** [comment]
Thanks .. LGTM

**Review by yeya24 — APPROVED** [review]

**Comment by damnever** [comment]
I am a little concerned about this change.

Context canceled generally means timeout happened then Prometheus(agent) or other clients may retry on this kind of situation. So the underlying ingesters may be overloaded because of this behavior since the timeout does not valid anymore.

**Comment by alanprot** [comment]
This change does not change the behavior at all.. 

Before and after we were creating a new background context to call the ingesters, and the reason for that is that we return as soon as we get the response from the quorum but we still wanna send the samples to all ingesters...

Am i missing something?

https://github.com/cortexproject/cortex/pull/5012/files#diff-8e51742acf8f88fa2ae01855fb19125feb5c54c41023f1eaa8beded9f1b4bcc3L787

**Pull request merged**